### ISSUE TITLE
PNG file stored in the flash of microcontroller

### ISSUE THREAD

**Issue by ali-rostami** [lvgl/lvgl#743]
My board is stm32f429 discovery and I need to display a png image which is stored in the flash memory of the microcontroller. After reading "[Use PNG images in LittlevGL](https://blog.littlevgl.com/2018-10-05/png_converter)"   I tried to display a 30 * 30 pixels, png image.
I used the codes in "[PNG decoder for LittelvGL](https://github.com/littlevgl/blog/raw/master/assets/png_converter/png_decoder.zip)" and found out that I have to change some codes in lodepng.c file I made these changes:
```
static void* lodepng_malloc(size_t size)
{
#ifdef LODEPNG_MAX_ALLOC
	if(size > LODEPNG_MAX_ALLOC) return 0;
#endif
	// printf("%s - %zd\n", __FUNCTION__, size);
	return lv_mem_alloc(size); //malloc(size);
}

static void* lodepng_realloc(void* ptr, size_t new_size)
{
#ifdef LODEPNG_MAX_ALLOC
	if(new_size > LODEPNG_MAX_ALLOC) return 0;
#endif
	// printf("%s - %zd\n", __FUNCTION__, new_size);
	return lv_mem_realloc(ptr, new_size);//realloc(ptr, new_size);
}

static void lodepng_free(void* ptr)
{
	// printf("%s\n", __FUNCTION__);
	lv_mem_free(ptr);//free(ptr);
}
```
**afer this I succeded but with the swaped colors of red and blue. why?**
Inorder to solving this problem in ```convert_color_depth```  function routin in png_decoder.c, I changed
 ```c = LV_COLOR_MAKE(img_argb[i].red, img_argb[i].green, img_argb[i].blue);```
to
 ```c = LV_COLOR_MAKE(img_argb[i].blue, img_argb[i].green, img_argb[i].red);``` .
I also have to mention that I use freertos and these are my defines in lv_conf.h
```
#define LV_MEM_CUSTOM_INCLUDE "FreeRTOS.h"   /*Header for the dynamic memory function*/
#define LV_MEM_CUSTOM_ALLOC   pvPortMalloc       /*Wrapper to malloc*/
#define LV_MEM_CUSTOM_FREE    vPortFree         /*Wrapper to free*/
```
Now I want to display an 800 * 480 png image which needs 800 * 400 * 4 bytes for decoding and as you know this space is not available in the internal ram. 
So **how can I assign the bank 2 of external ram to this?** -bank 1 is used for LCD display buffer-

**Comment by embeddedt** [comment]
@ali-rostami I think you'll have to make sure that your `malloc` functions can allocate memory from non-contiguous RAM (i.e. both banks). Or you can put your heap on just the external RAM, but that might slow things down.

**Comment by embeddedt** [comment]
@ali-rostami As you can see by reading the code, `lodepng_malloc` calls `lv_mem_alloc`, which (depending on `LV_MEM_CUSTOM`) will either:

1. If `LV_MEM_CUSTOM == 0`, use our internal memory allocator (which works off an array with a size of `LV_MEM_SIZE).
2. If `LV_MEM_CUSTOM == 1`, call through to your custom FreeRTOS allocator.

I think `configTOTAL_HEAP_SIZE` is what you're looking for, but you'll have to make sure the heap is stored in external RAM.

**Comment by ali-rostami** [comment]
what do you think about this method for allocating memory?
Is it good to use the freertos allocation function istead of the lvgl ones?
Storing the heap in the external RAM doesn't reduce the speed of the program or doesn't effect the performance or any other things?

**Comment by embeddedt** [comment]
@ali-rostami 

> what do you think about this method for allocating memory?
> Is it good to use the freertos allocation function istead of the lvgl ones?

It doesn't really matter. I think our memory allocator is relatively simplistic so I'm pretty sure that doesn't change performance that much.
> Storing the heap in the external RAM doesn't reduce the speed of the program or doesn't effect the performance or any other things?

It will definitely be somewhat slower than external RAM, but it might not be noticeable enough to be a major issue. You're using STM32F4 so you don't have a cache to compensate for the slower memory access.

**Comment by ali-rostami** [comment]
thank you very much dear @embeddedt for your answer.
How can I just put the lodepng needed space in the external ram? Is it possible?

**Comment by kisvegabor** [comment]
@ali-rostami 
If you use LittlevGL's built-in `lv_mem_alloc` function (`LV_MEM_CUSTOM == 0`) then you can try to write an external memory address [here](https://github.com/littlevgl/lvgl/blob/master/lv_misc/lv_mem.c#L82).
This way LittlevGL will allocate to the external RAM. If it works we can add an option for it to lv_conf.h

**Comment by ali-rostami** [comment]
@kisvegabor  the SDRAM on stm32f429 discovery has 4 banks. The first bank is used for the LCD frame buffer so I assigned the start address of the second bank (0xD0200000) for work_mem : 
```work_mem = (uint8_t *) 0xD0200000;// work_mem_int;```
but it didn't work. where did I do wrong?
BTW, I defined this: ```#define LV_MEM_CUSTOM       0```

**Comment by kisvegabor** [comment]
Is it working without the PNG decoder?

**Comment by ali-rostami** [comment]
No, before it gets there, it sticks in ```lv_init``` and give hard fault. -attempt to switch to invalid state-
seems it accessed to a wrong address but when I write and read a word from 0xD0200000 -bank 2 address- it doesn't give a hard fault.

**Comment by kisvegabor** [comment]
The SRAM is initialized before `lv_init()`?

**Comment by ali-rostami** [comment]
@kisvegabor yes.
I undo this line: ```work_mem = (uint8_t *) work_mem_int;``` and set 0xD0200000 as the lcd buffer and it worked.
It means there is no problem with this address.
After this line in ```lv_init``` it gives a hard fault:
```lv_obj_set_style(top_layer, &lv_style_transp_fit);```

**Comment by kisvegabor** [comment]
Hm, I have no idea. Can you upload your project somewhere? I also have an F429 DISCO board and I'd like to try it.

**Comment by ali-rostami** [comment]
Of course, I don't have access right now, but I will upload it as soon as possible.
//******************
update:
[Here you can download the project.](https://www.dropbox.com/s/2hoixgxn30kl700/stm32f429_discovery_7inch_lcd.rar?dl=0).
PS: IDE is truestudio. -an eclipse base IDE-

**Comment by kisvegabor** [comment]
I didn't try it yet just quickly checked the source code and didn't find the initialization of BANK1 next to BANK2.

Anyway, BANK1 should be enough for both the framebuffer and lv_mem. So if the frame buffer is working in BANK1 then the lv_mem address can be at `0xD0100000`. It is working this way?

**Comment by ali-rostami** [comment]
@kisvegabor 

> I didn't try it yet just quickly checked the source code and didn't find the initialization of BANK1 next to BANK2.

No there is no need to do this. The init is OK. The problem was for the expansion board I connected to the discovery board. I think there was a small solder that connected one of FMC to ground. After removing this now it's OK and there is no hard fault anymore.

**Comment by kisvegabor** [comment]
Awesome news! I'm waiting for the feedback. 
As I mentioned above, if it works well we can add an option to `lv_conf.h` to define the memory address externally.

**Comment by ali-rostami** [comment]
It doesn't work with an 800 * 480 image but it works with 150 * 150 one.
I set the drag property of img true and after some dragging it will disappear and says "No data".
Seems every time it wants to refresh the img, it takes space from external ram again until it gets to the end of it.

**Comment by kisvegabor** [comment]
It seems there is a memory leak somewhere. You can test it by calling `lv_mem_monitor` at the relevant points of the png decoder functions.

If a Bank has the size of 0x200000 then it's "only" 2MB. An 800x480 image in 32 bit is 1,5 MB so it's very near. Lodepng allocates memory to read the whole file + for the decoded result and some other things. So I can imagine that one Bank is not enough.

**Comment by ali-rostami** [comment]
This is the places I added ```lv_mem_monitor``` :
```
static void* lodepng_malloc(size_t size)
{
#ifdef LODEPNG_MAX_ALLOC
	if(size > LODEPNG_MAX_ALLOC) return 0;
#endif
	// printf("%s - %zd\n", __FUNCTION__, size);
	void * result = lv_mem_alloc(size); //malloc(size);

	lv_mem_monitor_t mon;
	lv_mem_monitor(&mon);
	printf("M) used: %6d (%3d %%), frag: %3d %%, big free: %6d\n",
			(int)mon.total_size - mon.free_size,
			mon.used_pct,
			mon.frag_pct,
			(int)mon.free_biggest_size);

	return result;
}

static void* lodepng_realloc(void* ptr, size_t new_size)
{
#ifdef LODEPNG_MAX_ALLOC
	if(new_size > LODEPNG_MAX_ALLOC) return 0;
#endif
	// printf("%s - %zd\n", __FUNCTION__, new_size);
	void * result = lv_mem_realloc(ptr, new_size);//realloc(ptr, new_size);

	lv_mem_monitor_t mon;
	lv_mem_monitor(&mon);
	printf("R) used: %6d (%3d %%), frag: %3d %%, big free: %6d\n",
			(int)mon.total_size - mon.free_size,
			mon.used_pct,
			mon.frag_pct,
			(int)mon.free_biggest_size);
	return result;
}

static void lodepng_free(void* ptr)
{
	// printf("%s\n", __FUNCTION__);
	lv_mem_free(ptr);//free(ptr);
	lv_mem_defrag();

	lv_mem_monitor_t mon;
	lv_mem_monitor(&mon);
	printf("F) used: %6d (%3d %%), frag: %3d %%, big free: %6d\n",
			(int)mon.total_size - mon.free_size,
			mon.used_pct,
			mon.frag_pct,
			(int)mon.free_biggest_size);
}
```
In the console I see that the used space gradually is increased until there is no space left and so then I see no data. 
![image](https://user-images.githubusercontent.com/9710249/51669989-e5c90980-1fda-11e9-88d0-f554f917b76c.png)

why does this happens? should it decode the png every time I move the image? why It doesn't free the space?

**Comment by kisvegabor** [comment]
I tested the memory leak in PC simulator with `#define LV_MEM_SIZE    (3000U * 1024U) `

Here my modified png_decoder:
[png_decoder.zip](https://github.com/littlevgl/lvgl/files/2796051/png_decoder.zip)

And the test code:
```c
    png_decoder_init();

    LV_IMG_DECLARE(png_decoder_test);
    lv_obj_t *img = lv_img_create(lv_scr_act(), NULL);
    lv_img_set_src(img, &png_decoder_test);
    lv_obj_set_drag(img, true
```

This way it was working well for me

**Comment by ali-rostami** [comment]
Thanks, @kisvegabor 
Yeah, Now It's working:sunglasses: I have 2 questions:
1- My color depth is 16, On my LCD I see that the order of red and blue are changed. so I changed this line:
```c = LV_COLOR_MAKE(img_argb[i].red, img_argb[i].green, img_argb[i].blue);```
to this:
```c = LV_COLOR_MAKE(img_argb[i].blue, img_argb[i].green, img_argb[i].red);```
**why does it happen?**
2- Every time I move the img object the png source needs to be decoded? If yes can we decode the png file once and then every time it needs call it from the external ram? 
you know, now It's slow and I'm thinking It would be great if all menu icons get decoded at startup and then LVGL could use them every time it needs. what do you think? Is it possible?

**Comment by kisvegabor** [comment]
Glad to hear it works!

You can decompress the image with lodepng at startup, copy the resulted bitmap to SRAM and initialize an `lv_img_dsc_t` with it. 
The idea is described [here](https://blog.littlevgl.com/2018-10-05/png_converter#add-and-test-the-png-decoder-library-with-a-file).

**Comment by ali-rostami** [comment]
Ok I will try. And what do you think about my first question? 
> 1- My color depth is 16, On my LCD I see that the order of red and blue are changed. so I changed this line:
> `c = LV_COLOR_MAKE(img_argb[i].red, img_argb[i].green, img_argb[i].blue);`
> to this:
> `c = LV_COLOR_MAKE(img_argb[i].blue, img_argb[i].green, img_argb[i].red);`
> **why does it happen?**

**Comment by kisvegabor** [comment]
I don't know why it is required. I would be possible to investigate where the color format changes but I'm not sure it's worth it if it works by just swapping red and blue.

**Comment by ali-rostami** [comment]
OK, I just worried if I missed something important.
Anyway, thanks @kisvegabor  and @embeddedt, :+1: Now It works perfectly.
 I defined a global var named ```my_test_png``` and added these lines at startup:
```
	my_test_png.header.always_zero = 0;
	my_test_png.header.w = 150;
	my_test_png.header.h = 150;
	my_test_png.header.cf = LV_IMG_CF_TRUE_COLOR_ALPHA;
	my_test_png.data_size = 22500 * LV_IMG_PX_SIZE_ALPHA_BYTE,
	my_test_png.data = decoder_open(&test_png, NULL);
```
then every where I can use my_test_png with high speed without bothering lag. 
I don't need external NAND flash for storing icons anymore and the internal flash of the microcontroller is enough. 😄

**Comment by kisvegabor** [comment]
Awesome!  :) 

It seems the issue is solved so if all clear, please close the issue!

**Comment by manison** [comment]
@ali-rostami 
> On my LCD I see that the order of red and blue are changed.

I had the exact same problem. I also solved it by swapping the red and blue colors. I didn't investigate yet whether the issue is within LittlevGL, LodePNG or whether both just use different color formats:

From [LodePNG FAQ](https://lodev.org/lodepng/)
> LodePNG uses raw buffers with RGBA or RGB pixels in the common order RGBARGBARGBA... or RGBRGBRGBRGB...
> ...
> Instead of using the order red, green, blue, it uses the order blue, green, red, or BGRBGRBGR..., (can result in wrong colors).

**Comment by ali-rostami** [comment]
OK.
 @kisvegabor just for reminding: please add an option to lv_conf.h to define the memory address externally. thanks

**Comment by ali-rostami** [comment]
oops, reading the decoded image from external ram is slower than reading it -true color with alpha- from the internal flash of the microcontroller.
It shows itself when I scroll the list. there is a big lag when list wants to refresh itself. -my list consists of imgbtn objects and it's not exactly the same as list object of lvgl-
**Any suggestion?**
It refreshes the list icons in 2 times, doesn't it? I'm saying this because I see the lag is exactly from the middle of icons, in other words, it first draws the first half of icons and after a little delay, draws the next half. Can I eliminate this delay?

**Comment by kisvegabor** [comment]
> @kisvegabor just for reminding: please add an option to lv_conf.h to define the memory address externally. thanks

Added as `LV_MEM_ADR` in `dev-5.3`.

> oops, reading the decoded image from external ram is slower than reading it -true color with alpha- from the internal flash of the microcontroller.

Yes the SRAM is slower than the internal flash.  Its clock is about 60 MHz (not sure). To read a data its address needs to be set and the data need to be read out.  I'm not sure how many cycles are really needed but if only 2 then you can read with 30 MHz. It's 1/6 of the flash speed in case of STMF429. To read 800x480 pixel it required 12 ms.

> It refreshes the list icons in 2 times, doesn't it? I'm saying this because I see the lag is exactly from the middle of icons, in other words, it first draws the first half of icons and after a little delay, draws the next half. Can I eliminate this delay?

It's because of the VDB size.  (`LV_VDB_SIZE` in lv_conf.h) increase it to let LittlevGL refresh bigger parts at once.

**Comment by ali-rostami** [comment]
yes, after increasing the VDB size it got better, although I couldn't double it because of the lack of internal ram.
The delay between filling the next VDB, depends on the rate calling of LV_Task? In other words, if I reduce the period of calling LV_Task to 1 millisecond the lag of drawing icons in the list should be reduced?

**Comment by ali-rostami** [comment]
>Have a look at this new feature: https://github.com/littlevgl/lvgl/blob/dev-5.3/lv_conf_templ.h#L91
You can move the VDB to the SRAM too with LV_VDB_ADR and LV_VDB_ADR2.

Just for making sure that I'm not missing anything, These are the changes I made:
```
#define LV_VDB_SIZE        LV_VER_RES * LV_HOR_RES
#define LV_VDB_ADR          0xD0000000
#define LV_VDB_DOUBLE       1
#define LV_VDB2_ADR         0xD0600000
#define LV_VDB_TRUE_DOUBLE_BUFFERED 1
```
and this is my ```ex_disp_flush``` function:
```
static void ex_disp_flush(int32_t x1, int32_t y1, int32_t x2, int32_t y2, const lv_color_t * color_p)
{
	if(color_p == LV_VDB_ADR)
		HAL_LTDC_SetAddress_NoReload(&hltdc, LV_VDB_ADR, LTDC_LAYER_1);
	else if(color_p == LV_VDB2_ADR)
		HAL_LTDC_SetAddress_NoReload(&hltdc, LV_VDB2_ADR, LTDC_LAYER_1);

	HAL_LTDC_Reload(&hltdc, LTDC_RELOAD_VERTICAL_BLANKING);
}
```
also defined this:
```
void HAL_LTDC_ReloadEventCallback(LTDC_HandleTypeDef *hltdc)
{
	lv_flush_ready();
}
```
Now it works and when I scroll the list there is no lag but It's so slow. I enabled the gpu -dma2d- It got better but still slow. the respond time of library seems reduced so much by this.
So what do you think?

**Comment by embeddedt** [comment]
@ali-rostami I'm going to make the educated guess that this is your SDRAM's limited speed. Unfortunately, without cache, I'm not sure there's much you can do about that.

I'm not sure what color depth you are using. Can you get by with 16-bit color instead of 32-bit color? It would use half the amount of RAM for everything.

**Comment by ali-rostami** [comment]
@embeddedt Hmm I hope it's not 😢 
I'm using 16-bit color.

**Comment by kisvegabor** [comment]
I agree with @embeddedt. I think that's the limit. Can you show a video about the speed?

**Comment by embeddedt** [comment]
@ali-rostami You mentioned at the beginning that your PNG takes up `800 * 480 * 4` bytes. Does that mean it's a 32-bit PNG? If so, you might as well use a 16-bit image, since your display won't show the extra color.

**Comment by ali-rostami** [comment]
@kisvegabor I will try.
@embeddedt No, If you take a look at ```decoder_open``` function of png_decoder.c, you will see that it uses ```lodepng_decode32``` of the lodepng library. so that's why I needed this amount of bytes to show a ```800 * 480``` png picture.

**Comment by ali-rostami** [comment]
I captured 2 times and you can get download them from [here.](https://www.dropbox.com/s/wzhmriysyo4eca5/vdb.rar?dl=0)
As you can see in **vdb_internal_ram.mp4**, there is a lag when I scroll the list at the bottom of icons. but the response time is acceptable and it reacts to the actions faster. 
By the way, I tried to make it looks better by reducing the width of the list and that's why I eliminated the gray background from the list here.
In **vdb_external_ram.mp4** you can see that there is no lag although the response time is low.

**Comment by kisvegabor** [comment]
Some questions:
- What is the VDB size in case of  "VDB internal RAM"? 
- Is it big enough to store the list?
- Have you enabled compiler optimization?
- So these icons are loaded into external SRAM?

**Comment by kisvegabor** [comment]
I see, thank you for the details.

There is one more thing you can try: Enable double buffering but not true double buffering.
1. Enable `LV_VDB_DOUBLE`
2. PLave the VDBs into internal RAM
3. Set a smaller VDB size because  two VDBs will be allocated 
4. In `disp_flush` use DMA to transfer `color_p` and when finished, in the interrupt call `lv_flush_ready()`

This you will have the speed of the internal RAM and the drawing and sending of the data become parallel.

**Comment by ali-rostami** [comment]
oh, Thanks for your suggestion but I tested it before but It wasn't good because the lag was shown in 2 places when I scrolled the list.
when I used the external ram at first I didn't use GPU -dma2d of stm32f429- so it was so slow, I mean so!
after enabling the GPU It got better. I could distinguish the differences easily.
But in the case of "VDB internal RAM", I can't distinguish the differences. how can I measure or make sure that using the GPU, in this case, have an impact on drawing speed? 
PS: Enabling GPU takes 1.5 KB more ram space.
